HA access to `kube-apiserver` using a host-local proxy

NicolasT opened the pull request:
Implement the 'local `nginx`' approach as described in #2103.

Fixes: #2103

NicolasT committed:
buildchain: distribute `nginx` as tar *and* in registry

Since we'll start using `nginx` as a 'normal' container, we need to
distribute it as both a tar archive (to inject in the `containerd` cache
on the bootstrap node), as well as in the 'registry' for clients to
pull.

See: https://github.com/scality/metalk8s/issues/2103

NicolasT committed:
kubernetes: add `127.0.0.1` to SAN of `kube-apiserver` cert

When using a local(host) `nginx` to proxy to the `kube-apiserver`
instances using a simple *stream* transport (i.e., TCP socket, no TLS
handling by `nginx`), we need to ensure the TLS server exposes a
certificate the client will accept, i.e., has `IP:127.0.0.1` in the
`subjectAlternateName` field.

See: https://github.com/scality/metalk8s/issues/2103

NicolasT committed:
salt: deploy a local proxy to `kube-apiserver` on all nodes

This is similar to the poor-mans-HA functionality as found in the
Kubespray project.

See: https://github.com/scality/metalk8s/issues/2103

NicolasT committed:
kubernetes: use control-plane IP in generated `KubeConfig`

Instead of using `BootstrapConfiguration` `apiServer.host`, which will
be removed in a subsequent commit since no longer required, use the
control-plane IP of the host on which the `KubeConfig` file is
generated.

See: https://github.com/scality/metalk8s/issues/2103

NicolasT committed:
salt, kubernetes: use the `apiserver-proxy` to connect to `kube-apiserver`

Instead of relying on the `BootstrapConfiguration` `apiServer.host`
value, we can instead use the proxy to `kube-apiserver` running on every
host in the cluster.

See: https://github.com/scality/metalk8s/issues/2103

NicolasT committed:
salt, kubernetes: remove `apiServer` from `BootstrapConfiguration`

We no longer need this since we provide in-cluster HA for
`kube-apiserver` access. If this is desired for out-of-cluster access,
we can provide this using a `LoadBalancer` `Service` once we have the
infrastructure to support this in place.

This also removed the optional deployment of `keepalived`.

See: https://github.com/scality/metalk8s/issues/2103
See: https://github.com/scality/metalk8s/issues/1788

bert-e commented:
# Hello nicolast,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/GH-2103-apiserver-ha` with contents from `improvement/GH-2103-apiserver-ha`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/GH-2103-apiserver-ha origin/development/2.5
 $ git merge origin/improvement/GH-2103-apiserver-ha
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/GH-2103-apiserver-ha
```

NicolasT commented:
There's a leftover reference to the `metalk8s:api_server:host` Pillar value in `salt/metalk8s/orchestrate/solutions/available.sls`, however, I'm not sure that's *supposed* to be there.

Need input @gdemonet.

NicolasT committed:
salt: use consistent permissions on SA private key

Using two different permissions for the same file doesn't really help
trying to make states idempotent.

NicolasT commented:
CI failures because in the upgrade/downgrade scenarios, `apiServer` config is still expected in `BootstrapConfiguration` (of course). Will fix.

NicolasT committed:
images: don't use `etcd3` 0.11.0 in `salt-master`

Install of this version is broken (missing `requirements/base.txt` file
used by `setup.py` in the package).

See: https://github.com/kragniz/python-etcd3/issues/952

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet commented on salt/metalk8s/kubernetes/apiserver-proxy/files/apiserver-proxy.yaml.j2:
```diff
@@ -0,0 +1,59 @@
+  name: apiserver-proxy
+  namespace: kube-system
+  labels:
+    addonmanager.kubernetes.io/mode: Reconcile
```
Oh, I was surprised by this "addon manager", looks like it is a "Bash Operator" :) https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/addon-manager/kube-addons.sh

slaperche-scality committed:
buildchain/image: allow to save image in multiple formats at once

We replace the `save_as_tar` boolean by a list of formats to use to save
the image.
This also allows to cleanup the code of RemoteImage by deporting
format-specific code into dedicated classes instead of branching hither
and yon in RemoteImage.

Refs: #2103
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>
(cherry picked from commit ac15e324be8b9deb186bfa805b0cdb841e6c5dbd)
Signed-off-by: Nicolas Trangez <ikke@nicolast.be>

slaperche-scality commented on buildchain/buildchain/image.py:
```diff
@@ -204,7 +207,9 @@ def _operator_image(name: str, **kwargs: Any) -> targets.OperatorImage:
-SAVE_AS_TAR : FrozenSet[str] = frozenset(('nginx', 'pause'))
+SAVE_AS_TAR : FrozenSet[str] = frozenset(['pause'])
+
+SAVE_AS_TAR_AND_IMAGE : FrozenSet[str] = frozenset(['nginx'])
```
I'm not very fond of this approach.

It's really ad-hoc/hackish, you can see it from the name `SAVE_AS_TAR_AND_IMAGE`.
Like if we ever add a third format, we will have `SAVE_AS_TAR_AND_IMAGE_AND_FOO` (unlikely I know, but it's still a code smell :wink: )

This aside, the current implementation breaks the cleanup: a `doit clean` on `_image_pull:nginx` won't delete the tarball :confused: 

I propose getting rid of the Boolean to use a list of formats instead. As a nice side effect it allows to have a simpler code in `RemoteImage` by moving the variant parts into dedicated class.
You can cherry-pick https://github.com/scality/metalk8s/commit/ac15e324be8b9deb186bfa805b0cdb841e6c5dbd if you want (should apply on top of your PR without conflict :slightly_smiling_face: ).

NicolasT replied:
Done

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

NicolasT committed:
salt: remove `metalk8s.internal.init`

This SLS really doesn't make much sense.

bert-e commented:
# History mismatch



Merge commit #2d6af2647b352daf7acee3aa6ee6f23d2771ceca on the integration branch
`w/2.5/improvement/GH-2103-apiserver-ha` is merging a branch which is neither the current
branch `improvement/GH-2103-apiserver-ha` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/GH-2103-apiserver-ha` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

NicolasT commented:
This seems to somehow break downgrade, because `etcd` doesn't get restarted. Further investigation shows this is due to the image not getting pulled?!?
Indeed, there's no `/var/lib/metalk8s/repositories/conf.d/99-metalk8s-2.3.0-dev-registry.inc` or similar on my test system. No clue what's causing that, though...

NicolasT committed:
scripts/iso-manager: override `metalk8s.api_server.kubeconfig` in Pillar

Previous MetalK8s states used during `iso-manager.sh -a` stage expect
the path to the `KubeConfig` to be used by `salt-master` to be available
in the Pillar (`metalk8s.api_server.kubeconfig`). This is removed in a
subsequent commit, replaced by a hard-coded value. As such, we need to
override this value (using the same hard-coded value) in the script,
otherwise importing e.g. a 2.3 ISO on a 2.4.x platform (containing these
changes) will fail.

Keeping this as a separate commit, mainly for documentation purposes.

See: https://github.com/scality/metalk8s/pull/2106

NicolasT committed:
ci: explicitly source `/etc/profile` when running build

For some reason `/etc/profile` isn't exported when `doit` is invoked.
As such, the build is not using the caching HTTP proxy we have in place
in our CI environment, resulting in a lot of unnecessary public traffic
or hitting upstream services.

Fixing by an explicit `source` of the environment file when invoking
`doit.sh`.

NicolasT commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/GH-2103-apiserver-ha` with contents from `improvement/GH-2103-apiserver-ha`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/GH-2103-apiserver-ha origin/development/2.5
 $ git merge origin/improvement/GH-2103-apiserver-ha
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/GH-2103-apiserver-ha
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality

NicolasT commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort





The following reviewers are expecting changes from the author, or must review again:

* @slaperche-scality







*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

gdemonet commented on salt/metalk8s/kubernetes/apiserver-proxy/files/apiserver-proxy.conf.j2:
```diff
@@ -0,0 +1,65 @@
+  upstream apiserver {
+    least_conn;
+{%- for apiserver in apiservers %}
+{%-     if apiserver == grains['metalk8s']['control_plane_ip'] %}
+{%-         set weight = 100 %}
+{%-      else %}
+{%-         set weight = 1 %}
+{%-      endif %}
```
FWIW, you can use [inline conditional expressions](https://jinja.palletsprojects.com/en/2.10.x/templates/#if-expression) with Jinja:

```
{%-     set weight = 100 if apiserver == grains['metalk8s']['control_plane_ip'] else 1 %}
```

TeddyAndrieux commented on salt/metalk8s/kubernetes/apiserver/installed.sls:
```diff
@@ -153,72 +93,14 @@ Create kube-apiserver Pod manifest:
-          - path: /etc/keepalived
-            name: keepalived-config
-{%- endif %}
         sidecars:
```
Maybe you should remove this key since it's empty

NicolasT replied:
Done

TeddyAndrieux commented on salt/metalk8s/kubernetes/apiserver/installed.sls:
```diff
@@ -153,72 +93,14 @@ Create kube-apiserver Pod manifest:
-              - name: keepalived-tmp
-                mountPath: /tmp
-{%- endif %}
         extra_volumes:
```
ditto

NicolasT replied:
Done

TeddyAndrieux commented on salt/metalk8s/kubernetes/apiserver/kubeconfig.sls:
```diff
@@ -4,7 +4,7 @@
   - metalk8s.internal.m2crypto
 
-{%- set apiserver = 'https://' ~ pillar.metalk8s.api_server.host ~ ':6443' %}
+{%- set apiserver = 'https://' ~ grains['metalk8s']['control_plane_ip'] ~ ':6443' %}
```
Why do not use the "proxy" here ?
That mean if for what ever reason local apiserver do not work the `admin.conf` will not work and since we use this kubeconfig everywhere it's not really good IMO

NicolasT replied:
Because this conf is perfectly valid to be used from *outside* the node, e.g. an admin copying it to his machine (aren't we even doing that for tests?)

Note how *most* `KubeConfig`s are rewritten to use the local proxy, so the services using those shouldn't be impacted. AFAIK only `salt-master` is using `admin.conf` (which is a bug by itself...).

In practice, not much changes, because in our current deployments the `apiServer.host` value is set to the hostname or IP of the bootstrap node anyway...

NicolasT replied:
Not that any *script* which uses `admin.conf` and is *certain* it's running on a host on which the proxy is deployed, could pass `-s`/`--server` to the `kubectl` invocation to use the proxy anyway.

TeddyAndrieux replied:
Ok but then I think we need to change something in our current restore script because if I well remember we need to contact apiserver before configuring the local one and we basically call `kubeconfig` salt state from apiserver and then use this new admin,conf

NicolasT replied:
That wouldn't work today in the first place then, unless the APIserver address was HA...

TeddyAndrieux replied:
Indeed 
```
.. warning::

    You cannot use the restore script if you do not have High Availability
    apiserver because some information required to reconfigure the others
    nodes are stored in the apiserver.
```
PS: Oh btw you should remove this note in this PR also

NicolasT replied:
No I mean, that's not a reasonable requirement in the first place.

Instead, the recover script should take, as an argument, the address of *an* API-server, and use that address when connecting to do whatever it needs to do.

TeddyAndrieux replied:
With this PR yes indeed but when restore script get writed in our bootstrap config the goal was to have an IP able to hit apiserver, and not only bound to bootstrap IP, either with keepalived or with any other system because configuring all server with the bootstrap IP as apiserver endpoint is really wrong.
But yes now we need to change this

NicolasT replied:
My point is that a VIP or LB for API-server has *always* been *optional*, so the restore script *relying* on that is wrong in the first place.

The bug fixed in this PR is the fact that sometimes we relied on this given address to be HA, given the context of #2108 

The restore script should also be updated to support this properly, and I'm happy to open a bug for this, but I'd rather not move this into this PR, since the context is completely different IMHO.

Finally, the lack of CI for the restore operation (#1687) basically *allows* me to merge this without remorse ;-P

NicolasT replied:
Ugh, I keep putting comments, but the topic keeps resonating.

What I was thinking, primarily: there *should* be no `admin.conf` in the backup, because it's not at all required to restore a cluster. An `admin.conf` can be generated from the CA data that *does* need to be backed up.

The whole restore process should be constructive, starting from the minimal set of required data as kept in a backup.

NicolasT replied:
So, I went to see, and apparently there's not even an `admin.conf` in the backup archive (which is good!)

However, what *is* in there has a really strange structure... We really need to spend some time on that it seems.

TeddyAndrieux commented on salt/metalk8s/orchestrate/upgrade/init.sls:
```diff
@@ -39,6 +39,15 @@ Wait for pillar refresh to complete:
 {%- set cp_nodes = salt.metalk8s.minions_by_role('master') | sort %}
 {%- set other_nodes = pillar.metalk8s.nodes.keys() | difference(cp_nodes) | sort %}
 
+Ensure all apiservers serve a certificate for 127.0.0.1:
+  salt.state:
+    - tgt: {{ cp_nodes | join(",") }}
+    - tgt_type: list
+    - sls:
+      - metalk8s.internal.upgrade.apiserver-cert-localhost
+    - saltenv: {{ saltenv }}
+    - batch: 1
```
I think this state should require `module: Wait for pillar refresh to complete`

NicolasT replied:
I'll do it, but could you explain why?

TeddyAndrieux replied:
Because we want the pillar to be refreshed before calling this, as you use information from the pillar and as just before this highstate we just upgrade the bootstrap our pillar may-be empty

TeddyAndrieux commented on salt/metalk8s/orchestrate/upgrade/init.sls:
```diff
@@ -53,9 +62,16 @@ Skip node {{ node }}, already in {{ node_version }} newer than {{ dest_version }
 
   {%- else %}
 
+Install apiserver-proxy on {{ node }}:
+  salt.state:
+    - tgt: {{ node }}
+    - sls:
+      - metalk8s.kubernetes.apiserver-proxy
+    - saltenv: {{ saltenv }}
```
This one should require `salt: Ensure all apiservers serve a certificate for 127.0.0.1`

NicolasT replied:
No, there's no dependency on that. `kubelet` reconfiguration requires the proxy to be up, and the cert to be resigned. But the proxy can happily be deployed and run, it doesn't really care about the services it's proxying to (at the TCP level).

NicolasT replied:
I added a `require` on `salt: Ensure all apiservers serve a certificate for 127.0.0.1` to `Deploy node {{ node }}` instead.

TeddyAndrieux commented on salt/metalk8s/orchestrate/upgrade/init.sls:
```diff
@@ -53,9 +62,16 @@ Skip node {{ node }}, already in {{ node_version }} newer than {{ dest_version }
+      - metalk8s.kubernetes.apiserver-proxy
+    - saltenv: {{ saltenv }}
+
 Wait for API server to be available on {{ node }}:
```
This one should require `salt: Install apiserver-proxy on {{ node }}` we don't want to run this one if proxy do not work otherwise we may wait for a timeout for ... nothing

NicolasT replied:
Agree,  done.

TeddyAndrieux commented on salt/metalk8s/salt/master/configured.sls:
```diff
@@ -13,10 +13,7 @@ Configure salt master:
-        {%- if pillar['metalk8s']['api_server'].get('context') %}
-        kubecontext: "{{ pillar['metalk8s']['api_server']['context'] }}"
-        {%- endif %}
+        kubeconfig: "/etc/kubernetes/admin.conf"
```
Like already said on slack for me this change should land in `dev/2.0` to not have to edit `scripts/iso-manager.sh` in this PR

NicolasT replied:
I'll do a separate PR to handle that. Don't want this one to go in 2.0, and don't want to wait for something to land in 2.0 which is primarily 'cosmetic'.

TeddyAndrieux reviewed (commented):
Looks good just few comments

bert-e commented:
# History mismatch



Merge commit #edb50cc539c1bad95a84d59ec3a8fd77e294bbcf on the integration branch
`w/2.5/improvement/GH-2103-apiserver-ha` is merging a branch which is neither the current
branch `improvement/GH-2103-apiserver-ha` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/GH-2103-apiserver-ha` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **approve**

NicolasT commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/GH-2103-apiserver-ha` with contents from `improvement/GH-2103-apiserver-ha`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/GH-2103-apiserver-ha origin/development/2.5
 $ git merge origin/improvement/GH-2103-apiserver-ha
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/GH-2103-apiserver-ha
```





*The following options are set:* **approve**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

TeddyAndrieux reviewed (approved):
LGTM (except restore script which is broken now but will be tackle in another PR)

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

gdemonet commented on CHANGELOG.md:
```diff
@@ -2,6 +2,15 @@
 
 ## Release 2.4.2 (in development)
 
+### Breaking changes
```
Would probably make sense to bump the version of `BootstrapConfiguration`

NicolasT replied:
No. What you can put in the 'object' doesn't really change: there are no backwards-incompatible changes or anything. We just now ignore information we did use before (and as such, no longer require them to be set). It's a required value that now becomes kinda 'optional' (and unused, at all).

Basically, any existing `v1alpha2` `BootstrapConfiguration` *still* works.

gdemonet reviewed (changes requested):
Works on my machine, and code looks good to me, except for the `apiVersion` of the `BootstrapConfiguration` which should change (as you documented it, it's a breaking change).

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue GH-2103.

Goodbye nicolast.
